apis/postgresql/v1alpha1/grant_types.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+)
+
+// A GrantSpec defines the desired state of a Grant.
+type GrantSpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+	ForProvider       GrantParameters `json:"forProvider"`
+}
+
+// GrantPrivilege represents a privilege to be granted
+// +kubebuilder:validation:Pattern:=[A-Z]+
+type GrantPrivilege string
+
+// If Privileges are specified, we should have at least one
+
+// GrantPrivileges is a list of the privileges to be granted
+// +kubebuilder:validation:MinItems:=1
+type GrantPrivileges []GrantPrivilege
+
+// ToStringSlice converts the slice of privileges to strings
+func (gp *GrantPrivileges) ToStringSlice() []string {
+	out := make([]string, len(*gp))
+	for i, v := range *gp {
+		out[i] = string(v)
+	}
+	return out
+}
+
+// GrantParameters define the desired state of a PostgreSQL grant instance.
+type GrantParameters struct {
+	// Privileges to be granted.
+	// See https://www.postgresql.org/docs/current/sql-grant.html for available privileges.
+	// +optional
+	Privileges GrantPrivileges `json:"privileges,omitempty"`
+
+	// WithGrantOption allows a granted role to grant the given privileges to another role.
+	// +optional
+	WithGrantOption *bool `json:"withGrantOption,omitempty"`
+
+	// WithAdminOption allows a granted role to grant another role membership of a group.
+	// +optional
+	WithAdminOption *bool `json:"withAdminOption,omitempty"`
```

Comment:
The type of option that can be used depends on the type of grant the user is creating, i.e. `WITH GRANT OPTION` only works when setting permissions on a database and `WITH ADMIN OPTION` only works when setting permissions on role membership.

In hindsight, 2 boolean options for this makes the representation more complex than it needs to be, but I don't think an array of options makes sense either since you can't actually use more than one of the options at once.

Maybe a single string field (`withOption`) with an `Enum` validator applied makes sense? Leaves the door open for additional options to be added in future (very unlikely at this point but who knows), avoids usage of booleans and avoids any internal logic deciding which option to set based on what privilege type the user wants. The downside there is if the user specifies an option that isn't compatible with their `Grant` then they'll get an SQL error.

Reply:
That sounds great - I'd definitely go with an enum given the options are mutually exclusive.